File: Data Structures & Algorithms/string-encode-and-decode/submission-4.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        encoded = ""
        for s in strs:
            str_len = len(s)
            encoded = encoded + f"{str_len}-"+s

        return encoded

    def decode(self, s: str) -> List[str]:
        ans = []
        i = 0
        while i < len(s):
            j = i
            while s[j] != "-":
                j+=1
            str_len = int(s[i:j])
            ans.append(s[j+1:j+str_len+1])
            i = j+1 + str_len
        return ans

File: Data Structures & Algorithms/anagram-groups/submission-4.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        // return a list with lists of strings, length of ans list is length of strs
        //can have arraylist inside of main List
        // need a word frequency map to order words
        // append if hash of frequ matches existant
        /*
        return type is list generic, can use arraylist
        create arraylist for ans
        create freqmap for saving hashes

        #hash creation
        for every char in the word saved the char in charcode ofset (index in array len26)
        call array to string to create inmutable for freq map key
        if key exists the append, else put new key and value
        for every key and value in the map append value to ans (value is array of valid anagrams)
        return ans
        */
        // step 1 create freqmap and ans list
        Map<String, List<String>> freqMap = new HashMap<>();
        ArrayList<List<String>> ans = new ArrayList<>();

        //create hashes
        for(String word: strs){
            int[] key = new int[26];
            for(char c : word.toCharArray()){
                int index = c - 'a';
                key[index]++;
            }
            String hashKey = Arrays.toString(key);
            freqMap.putIfAbsent(hashKey, new ArrayList<>());
            freqMap.get(hashKey).add(word);
        }

        for(String hkey : freqMap.keySet()){
            ans.add(freqMap.get(hkey));
        }
        return ans;
    }
}
